### DISCUSSION TITLE
Adding support for "partial records"

### DISCUSSION THREAD

**Discussion by ldeluigi** [riok/mapperly#417] [Ideas]
Hello, I was just wondering why this is not currently supported:
```csharp
[Mapper]
public partial record MembershipDto 
{
    public static partial MembershipDto ToDto(MembershipSnapshot src);
}
```

While this is:
```csharp
[Mapper]
public partial class MembershipDto 
{
    public partial MembershipDto ToDto(MembershipSnapshot src);
}
```

**Comment by latonz** [comment]
There was no use case / demand (yet). Why would you prefer a record over a class for a mapper?

**Reply by ldeluigi** [reply]
Because we are using records as Dto types and we'd like to generate mappers without the burden of creating ad-hoc (static) classes for every single one of them. This feature would also require to allow generating static partial method implementation for non static classes, if it's possible.

**Reply by CommonGuy** [reply]
Ah, you are declaring the mapper method directly in the same record that you want to map.

We understand the use case now. It's not a high priority for us, since this probably won't be used by many people.

**Reply by ldeluigi** [reply]
That's unfortunate but thanks for your replies

**Reply by TimothyMakkison** [reply]
> This feature would also require to allow generating static partial method implementation for non static classes, if it's possible.

This is supported! See #364.

You can create a static mapper class with extension methods for your records.

```C#
[Mapper(EnumMappingStrategy = EnumMappingStrategy.ByName)]
public static partial class Mapper
{
    [MapProperty(nameof(Car.Manufacturer), nameof(CarDto.Producer))] // Map property with a different name in the target type
    public static partial CarDto MapCarToDto(this Car car);
    public static partial DogDto MapDogToDto(this Dog value);
    public static partial CatDto MapCatToDto(this Cat value);
}
```

**Reply by ldeluigi** [reply]
Sorry, I wrote:
> This feature would also require to allow generating static partial method implementation for non static classes, if it's possible.

but I meant:
> This feature would also require to allow generating static partial method implementations **inside** non static classes, if it's possible.

Afaik, it's not currently supported to generate static partial mappers inside non static (dto) classes.

### More context
With the addition of "static abstract" methods in C#11 we can now define such interface:
```csharp
public interface IMappableFrom<F, T>
    where T : IMappableFrom<F, T>
{
    static abstract T MapFrom(F src);
}
```

This would allow dto records to implement the interface in order to be mapped by a generic
```csharp
public ActionResultBuilder<TResult, TNewDto, TMeta> MapTo<TNewDto>()
        where TNewDto : IMappableFrom<TDto, TNewDto> => Map(TNewDto.MapFrom);
```

Our desideratum would be to have the `static abstract T MapFrom(F src);` source generated by mapperly for records that implement the interface.
